plugins/preserve-tombs.cpp

```diff
@@ -0,0 +1,269 @@
+}
+
+static bool assign_to_tomb(int32_t unit_id, int32_t building_id);
+static void update_tomb_assignments(color_ostream& out);
+void onUnitDeath(color_ostream& out, void* ptr);
+static command_result do_command(color_ostream& out, std::vector<std::string>& params);
+
+DFhackCExport command_result plugin_init(color_ostream &out, std::vector <PluginCommand> &commands) {
+    commands.push_back(PluginCommand(
+        plugin_name,
+        "Preserves tomb assignments to units when they die.",
+        do_command));
+    return CR_OK;
+}
+
+static command_result do_command(color_ostream& out, std::vector<std::string>& params) {
+    if (params.size() != 1 || params[0] != "status") {
+        out.print("%s wrong usage", plugin_name);
+        return CR_WRONG_USAGE;
+    }
+    else {
+        out.print("%s is currently %s\n", plugin_name, is_enabled ? "enabled" : "disabled");
+        if (is_enabled) {
+            out.print("tracked tomb assignments:\n");
+            std::for_each(tomb_assignments.begin(), tomb_assignments.end(), [&out](const auto& p){
+                auto& [unit_id, building_id] = p;
+                out.print("unit %d -> building %d\n", unit_id, building_id);
+            });
+        }
+        return CR_OK;
+    }
+}
+
+// event listener
+EventManager::EventHandler assign_tomb_handler(onUnitDeath, 0);
+
+DFhackCExport command_result plugin_enable(color_ostream &out, bool enable) {
+    tomb_assignments.clear();
+    if (!Core::getInstance().isWorldLoaded()) {
+        out.printerr("Cannot enable %s without a loaded world.\n", plugin_name);
+        return CR_FAILURE;
+    }
+
+    if (enable != is_enabled) {
+        is_enabled = enable;
+        DEBUG(config,out).print("%s from the API; persisting\n",
+                                is_enabled ? "enabled" : "disabled");
+        set_config_bool(config, CONFIG_IS_ENABLED, is_enabled);
+        EventManager::registerListener(EventManager::EventType::UNIT_DEATH, assign_tomb_handler, plugin_self);
+        if (enable)
+            update_tomb_assignments(out);
+    } else {
+        EventManager::unregisterAll(plugin_self);
+        DEBUG(config,out).print("%s from the API, but already %s; no action\n",
+                                is_enabled ? "enabled" : "disabled",
+                                is_enabled ? "enabled" : "disabled");
+    }
+    return CR_OK;
+}
+
+DFhackCExport command_result plugin_shutdown (color_ostream &out) {
+    DEBUG(config,out).print("shutting down %s\n", plugin_name);
+
+    return CR_OK;
+}
+
+DFhackCExport command_result plugin_load_data (color_ostream &out) {
+    cycle_timestamp = 0;
+    config = World::GetPersistentData(CONFIG_KEY);
+
+    if (!config.isValid()) {
+        DEBUG(config,out).print("no config found in this save; initializing\n");
+        config = World::AddPersistentData(CONFIG_KEY);
+        set_config_bool(config, CONFIG_IS_ENABLED, is_enabled);
+        set_config_val(config, CONFIG_CYCLES, 25);
```

Comment:
this is even more frequent than `nestboxes`, which uses a super low cycle time of 50 ticks (half a second at 100 fps). I would think 100 is more appropriate here, which is once per second at 100 fps and once every 5 seconds at 20 fps. would that be ok?

Reply:
That probably would be fine yeah - I've adjusted the defaults